site-src/geps/gep-992.md

```diff
@@ -0,0 +1,184 @@
+- add the "least-connections" strategy
+
+## Non-Goals
+
+- defining combinations of strategies (e.g. "least-connections" + "weighted")
+
+## Introduction
+
+It's currently possible to configure multiple backends for an `HTTPRoute` rule, e.g.:
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1alpha2
+kind: HTTPRoute
+metadata:
+  name: multiple-backends
+spec:
+  rules:
+  - matches:
+    - path:
+        type: PathPrefix
+        value: /bar
+    backendRefs:
+    - name: service1
+      port: 80
+    - name: service2
+      port: 80
+    - name: service3
+      port: 80
+```
+
+This implies a load-balancing strategy for traffic which matches but is limited
+in how it can be configured today.
+
+### Problems
+
+Currently only one load-balancing strategy is available, which we'll call the
+`weighted` strategy (and by proxy, you can implement a `round-robin` strategy
+by providing equal weights to each backend).
+
+Problematically, this default `weighted` strategy is implied by means of
+setting weights on individual objects, and consequently there is no place to
+declare an intended state for other strategies.
+
+It is not clear (via specification) to an end-user what the load-balancing
+implementation for any given `HTTPRoute` will end up being when they post
+an new `HTTPRoute` object.
+
+### Proposed Solution
+
+Route rules can now be provided a `LoadBalancerStrategy`, which is required
+but defaults to the generic `RoundRobinStrategy`:
+
+```golang
+type HTTPRouteRule struct {
+	// <NO documentation updates>
+	Matches []HTTPRouteMatch `json:"matches,omitempty"`
+
+	// <NO documentation updates>
+	Filters []HTTPRouteFilter `json:"filters,omitempty"`
+
+	// BackendRefs defines references to the individual backend(s) where matching
+	// requests should be sent.
+	//
+	// If multiple backends are configured the LoadBalancerStrategy option becomes
+	// required and will define how traffic will be dispersed across those backends.
+	//
+	// If unspecified or invalid (refers to a non-existent resource or a Service
+	// with no endpoints), the rule performs no forwarding. If there are also no
+	// filters specified that would result in a response being sent, a HTTP 503
+	// status code is returned.
+	//
+	// Support: Core for Kubernetes Service
+	// Support: Custom for any other resource
+	//
+	// Support for weight: Core
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	BackendRefs []HTTPBackendRef `json:"backendRefs,omitempty"`
+
+	// LoadBalancerStrategy defines the high level strategy to use when multiple
+	// BackendRefs are configured for traffic.
+	//
+	// If only one backend is provided in BackendRefs this property has no
+	// effect, otherwise it is required.
+	//
+	// Support: Core for RoundRobinStrategy
+	// Support: Core for WeightedStrategy
+	// Support: Extended for LeastConnectionsStrategy
+	//
+	// +optional
+	// +kubebuilder:validation:Enum=RoundRobin,WeightedStrategy,LeastConnectionsStrategy
```

Comment:
Reading through this GEP got me to spend some time looking at what was possible across different implementations. Similar to the rewrite GEP, I think it would be useful to provide a brief overview of the capabilities of various proxies/load balancers to clarify why we're choosing theses strategies as a base. Here's a rough starting point for what I've found so far. 

**GCP supports [3 modes:](https://cloud.google.com/load-balancing/docs/backend-service#balancing-mode)**
- Connection
- Rate 
- Utilization

**Envoy supports [5 balancing modes](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/load_balancers#arch-overview-load-balancing-types-least-request):**
* Weighted round robin
* Weighted least request
* Ring hash
* Maglev
* Random

**NGINX supports [6 balancing modes:](https://docs.nginx.com/nginx/admin-guide/load-balancer/http-load-balancer/):**
* Round Robin
* Least Connections
* IP Hash
* Generic Hash
* Least Time
* Random

**HAPROXY supports [10 balancing modes:](http://cbonte.github.io/haproxy-dconv/2.0/configuration.html#4-balance)**
* roundrobin
* static-rr
* leastconn
* first
* source
* uri
* url_param
* hdr
* random
* rdp-cookie


Some top level concerns:
- Different proxies support attaching this config at different levels
- They almost universally accept additional configuration with varying structure

I need to think about this more, but I'm wondering if this could be better served with policy attachment.

Reply:
I agree with this -- I think this is something that is going to be hard to bridge in a good way, especially between distributed vs single instance implementations.

Reply:
As mentioned in the other comments, it is common to have two levels of load balancing:
- first, choose a backend (load balancing pool, upstream, cluster, etc.)
- second, choose a backend instance according to the load balancing method.

I don't think it is common for proxies to support least connections or round robin methods at the first level expect splitting traffic by weight . As a result, to implement the proposed configuration, it will be necessary to create a new pool to include members from different pools. 

My concern is that we will remove the first level this way. And as mentioned here https://github.com/kubernetes-sigs/gateway-api/pull/993/files#r784229605, it will be necessary to create a pool per each route.